### DISCUSSION TITLE
Accessing the PDF StructTreeRoot

### DISCUSSION THREAD

**Discussion by cole-dda** [pymupdf/PyMuPDF#2389] [Q&A]
**Is your feature request related to a problem? Please describe.**
i want access pdf's structtreeroot,but i search document api,not idea how get it

**Describe the solution you'd like**
type,value = doc.xref_object(xref)
value is str

i want:
type,value = doc.xref_object(xref,convert=True)
value is python object,such as,dict list str bool None

**Comment by JorjMcKie** [comment]
Unfortunately, your description is very vague about what your requirement is, what you are trying to achieve ultimately and specificaclly, what the results should be e.g., what is the meaning of the "convert" parameter?
But as a first cut, here is what you can do already:

You can use `Document` methods `xref_get_key()`, `xref_get_keys()` and `pdf_catalog()` to find the `/StructTreeRoot` object and its content. This works like this:
1. Find the xref of the catalog: `catxref = doc.pdf_catalog()`
2. Access the `/StructTreeRoot` dictionary : `structroot = doc.xref_get_key(catxref, "StructTreeRoot")`

The result of the second method always is a tuple of strings - see documentation.
Depending on how your PDF is made, three responses are possible:
1. `("null", "null")` - this PDF has no StructTreeRoot.
2. `("xref", "n 0 R")` - a StructTreeRoot is present and can be found at xref `n`. Access it via `doc.xref_object(n)`.
3. `("dict", "<< ... >>")` - a StructTreeRoot is present and given as the second item.

Before we can talk about an actual enhancement, a lot of clarification work is necessary obviously.
I suggest to do this under the tab "Discussions", and I will transfer this item to there.
